### PR TITLE
feat: asyncio with retries

### PR DESCRIPTION
**Pull request by apinanyogaratnam**
This pull request introduces and tests new features for the `tool_call` decorator, including support for asynchronous execution, retry logic, timeout and backoff parameters, and concurrent execution. The changes ensure that these new parameters are validated, their behaviors are correct, and that schema generation and validation remain consistent. The test suites have been expanded to cover these features, and the decorator implementation has been updated to support them.

**Decorator enhancements and validation:**

* Added new parameters to `tool_call`: `retry`, `timeout`, `max_backoff`, and `concurrent`, with validation to ensure `retry` requires `raise_for_exception` and sensible defaults for timeouts and backoff.
* Improved schema generation logic to maintain consistency and proper handling of required and optional parameters with new decorator features.

**Testing async, retry, and concurrent features:**

* Added comprehensive tests in `tests/test_decorators.py` for async execution (`run_async`), retry logic (including eventual success and failure), concurrent execution, and validation of new parameters.
* Expanded `tests/test_registry.py` to test async tool execution via `Toolkit.run_tool_async`, including error handling, concurrent execution, retry-enabled tools, and validation errors.

**Validation and schema consistency:**

* Added tests in `tests/test_validation.py` to ensure schema generation and validation error messages remain correct with new decorator features, including edge cases and complex types.

**Imports for async and concurrency:**

* Updated imports in relevant test and implementation files to include `asyncio` and `concurrent.futures` for supporting new async and concurrent features. [[1]](diffhunk://#diff-de499b913ec172e2c2928e4a3615f8985e06e6fdde95b85be3218d827ae28d58R4-R6) [[2]](diffhunk://#diff-17d5ee9ae3f62cf038b21b07d665043c0f4990bf61b153015121621d6852dbc1R2-R3) [[3]](diffhunk://#diff-0abc741bfcebe9debcf5eb70d01e66db1ccb9036877f195680d25bf73e9e9083R2) [[4]](diffhunk://#diff-f1dbe79adfb6e39e76ee4519368bc4b5eaa0e4831f00165e5ac28fede4825bffL2-R3)

These changes collectively enhance the flexibility and robustness of the `tool_call` decorator and its integration in the toolkit, ensuring reliable execution and validation in both synchronous and asynchronous contexts.

### CODE REVIEW

File: vaul/decorators.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 from __future__ import annotations
 
 import inspect
+import time
+import asyncio
+import concurrent.futures
```
Comment (4thabang) [inline vaul/decorators.py:6]:
```suggestion
from concurrent.futures import ThreadPoolExecutor
```

**Review by 4thabang — CHANGES_REQUESTED** [review]
Left a few changes and comments, haven't reviewed tests just yet. Solid PR, all thats missing now is documentation.
